forte/data/data_store.py

```diff
@@ -892,27 +911,93 @@ def co_iterator_annotation_like(
+            )
+
+        entry_class = get_class(type_name)
+        all_types = set()
         if include_sub_type:
-            entry_class = get_class(type_name)
-            all_types = []
-            # iterate all classes to find subclasses
             for type in self.__elements:
                 if issubclass(get_class(type), entry_class):
-                    all_types.append(type)
+                    all_types.add(type)
+        else:
+            all_types.add(type_name)
+        all_types = list(all_types)
+        all_types.sort()
+        if self._is_annotation(type_name):
+            if range_annotation is None:
+                yield from self.co_iterator_annotation_like(all_types)
+            else:
+                for entry in self.co_iterator_annotation_like(all_types):
+                    if within_range(entry, range_annotation):
+                        yield entry
+        elif issubclass(entry_class, Link):
             for type in all_types:
-                yield from self.iter(type)
+                if range_annotation is None:
+                    yield from self.iter(type)
+                else:
+                    for entry in self.__elements[type]:
+                        if (
+                            entry[constants.BEGIN_INDEX] in self.__tid_ref_dict
+                        ) and (
+                            entry[constants.END_INDEX] in self.__tid_ref_dict
+                        ):
+                            parent = self.__tid_ref_dict[
+                                entry[constants.BEGIN_INDEX]
+                            ]
+                            child = self.__tid_ref_dict[
+                                entry[constants.END_INDEX]
+                            ]
+                            if within_range(
+                                parent, range_annotation
+                            ) and within_range(child, range_annotation):
+                                yield entry
+        elif issubclass(entry_class, Group):
+            for type in all_types:
+                if range_annotation is None:
+                    yield from self.iter(type)
+                else:
+                    for entry in self.__elements[type]:
+                        member_type = entry[constants.BEGIN_INDEX]
+                        if self._is_annotation(member_type):
+                            members = entry[constants.END_INDEX]
+                            within = True
+                            for m in members:
+                                e = self.__tid_ref_dict[m]
+                                if not within_range(e, range_annotation):
+                                    within = False
+                                    break
+                            if within:
+                                yield entry
         else:
             if type_name not in self.__elements:
                 raise KeyError(f"type {type_name} does not exist")
```

Comment:
1. we should use check this at the very beginning without going through all the other steps
   - we probably should use `ValueError` since this is an incorrect argument provided from the function signature.
2. what happens if this item is not `Group`, `List`, or `Annotation`?  Maybe we should say we cannot handle.

Reply:
yes